File: packages/template-go/loader/loader.go
// Package loader defines template loaders and name resolution (RT-7 to RT-10).
package loader

import (
	"errors"
	"fmt"
	"hash/fnv"
	"io/fs"
	"strings"

	"github.com/polyspec/template/ast"
)

// Result is a loaded template: source text or a parsed AST, and a version.
type Result struct {
	Source  []byte
	AST     *ast.Template
	Version string
}

// Loader returns the template for a name; ok is false when the name does not exist.
type Loader interface {
	Load(name string) (Result, bool)
}

// ContentHash returns the FNV-1a hash of bytes as hex.
func ContentHash(data []byte) string {
	h := fnv.New32a()
	h.Write(data)
	return fmt.Sprintf("%08x", h.Sum32())
}

// MapLoader holds template sources in memory.
type MapLoader struct {
	entries map[string]Result
}

// NewMapLoader creates a map loader from name → source text.
func NewMapLoader(sources map[string]string) *MapLoader {
	l := &MapLoader{entries: map[string]Result{}}
	for name, source := range sources {
		l.Set(name, source)
	}
	return l
}

// Set stores a source.
func (l *MapLoader) Set(name, source string) {
	l.entries[name] = Result{Source: []byte(source), Version: ContentHash([]byte(source))}
}

// SetAST stores a parsed template.
func (l *MapLoader) SetAST(name string, template *ast.Template) {
	l.entries[name] = Result{AST: template, Version: ContentHash([]byte(name))}
}

// Load implements Loader.
func (l *MapLoader) Load(name string) (Result, bool) {
	r, ok := l.entries[name]
	return r, ok
}

// FSLoader reads templates from a file system.
type FSLoader struct {
	fsys fs.FS
}

// NewFSLoader creates a loader over a file system rooted at the loader root.
func NewFSLoader(fsys fs.FS) *FSLoader {
	return &FSLoader{fsys: fsys}
}

// Load implements Loader; the version is the modification time and size.
func (l *FSLoader) Load(name string) (Result, bool) {
	if !fs.ValidPath(name) {
		return Result{}, false
	}
	info, err := fs.Stat(l.fsys, name)
	if err != nil || info.IsDir() {
		return Result{}, false
	}
	data, err := fs.ReadFile(l.fsys, name)
	if err != nil {
		return Result{}, false
	}
	return Result{Source: data, Version: fmt.Sprintf("%d:%d", info.ModTime().UnixNano(), info.Size())}, true
}

// ErrOutsideRoot is returned when a path leaves the loader root.
var ErrOutsideRoot = errors.New("path leaves the loader root")

// ResolvePath implements RT-8.
func ResolvePath(current, path string) (string, error) {
	var segments []string
	if !strings.HasPrefix(path, "/") {
		parts := strings.Split(current, "/")
		for _, part := range parts[:len(parts)-1] {
			if part != "" {
				segments = append(segments, part)
			}
		}
	}
	for _, segment := range strings.Split(path, "/") {
		switch segment {
		case "", ".":
			continue
		case "..":
			if len(segments) == 0 {
				return "", fmt.Errorf("%q: %w", path, ErrOutsideRoot)
			}
			segments = segments[:len(segments)-1]
		default:
			segments = append(segments, segment)
		}
	}
	return strings.Join(segments, "/"), nil
}

File: packages/template-rust/src/loader.rs
//! Loader trait, map loader, filesystem loader and template name resolution (RT-7 to RT-10).

use crate::ast::Template;
use std::collections::HashMap;
use std::path::{Path, PathBuf};

/// A loaded template: source bytes or a parsed AST, with a version.
#[derive(Debug, Clone)]
pub enum Loaded {
    /// Source text.
    Source {
        /// UTF-8 bytes.
        bytes: Vec<u8>,
        /// Version of the source.
        version: String,
    },
    /// A parsed template.
    Ast {
        /// The AST.
        ast: Template,
        /// Version of the AST.
        version: String,
    },
}

impl Loaded {
    /// The version.
    pub fn version(&self) -> &str {
        match self {
            Loaded::Source { version, .. } | Loaded::Ast { version, .. } => version,
        }
    }
}

/// Loads templates by name (RT-9).
pub trait Loader {
    /// Returns the template for a name, or None when the name does not exist.
    fn load(&self, name: &str) -> Option<Loaded>;
}

/// FNV-1a hash of bytes, used as the version of in-memory sources.
pub fn content_hash(bytes: &[u8]) -> String {
    let mut hash: u32 = 0x811c9dc5;
    for byte in bytes {
        hash ^= *byte as u32;
        hash = hash.wrapping_mul(0x01000193);
    }
    format!("{hash:08x}")
}

/// In-memory loader.
#[derive(Debug, Default)]
pub struct MapLoader {
    entries: HashMap<String, Loaded>,
}

impl MapLoader {
    /// Creates an empty loader.
    pub fn new() -> MapLoader {
        MapLoader::default()
    }

    /// Adds a source.
    pub fn set(&mut self, name: &str, source: &str) {
        self.entries.insert(
            name.to_string(),
            Loaded::Source {
                bytes: source.as_bytes().to_vec(),
                version: content_hash(source.as_bytes()),
            },
        );
    }

    /// Adds a parsed template.
    pub fn set_ast(&mut self, name: &str, ast: Template) {
        let version = content_hash(serde_json::to_string(&ast).unwrap_or_default().as_bytes());
        self.entries.insert(name.to_string(), Loaded::Ast { ast, version });
    }
}

impl Loader for MapLoader {
    fn load(&self, name: &str) -> Option<Loaded> {
        self.entries.get(name).cloned()
    }
}

/// Filesystem loader rooted at a directory (RT-10).
#[derive(Debug, Clone)]
pub struct FsLoader {
    root: PathBuf,
}

impl FsLoader {
    /// Creates a loader for a root directory.
    pub fn new(root: impl AsRef<Path>) -> FsLoader {
        FsLoader {
            root: root.as_ref().to_path_buf(),
        }
    }

    /// Path of a template name under the root.
    pub fn path_of(&self, name: &str) -> PathBuf {
        let mut path = self.root.clone();
        for segment in name.split('/') {
            path.push(segment);
        }
        path
    }
}

impl Loader for FsLoader {
    fn load(&self, name: &str) -> Option<Loaded> {
        let path = self.path_of(name);
        let metadata = std::fs::metadata(&path).ok()?;
        if !metadata.is_file() {
            return None;
        }
        let bytes = std::fs::read(&path).ok()?;
        let modified = metadata
            .modified()
            .ok()
            .and_then(|time| time.duration_since(std::time::UNIX_EPOCH).ok())
            .map(|d| d.as_nanos())
            .unwrap_or(0);
        Some(Loaded::Source {
            bytes,
            version: format!("{modified}:{}", metadata.len()),
        })
    }
}

/// Error of a path that leaves the loader root.
#[derive(Debug)]
pub struct PathError(pub String);

/// RT-8: resolves a path written in a tag against the directory of the current template.
pub fn resolve_path(current: &str, path: &str) -> Result<String, PathError> {
    let mut segments: Vec<&str> = if path.starts_with('/') {
        Vec::new()
    } else {
        let parts: Vec<&str> = current.split('/').collect();
        parts[..parts.len().saturating_sub(1)]
            .iter()
            .copied()
            .filter(|s| !s.is_empty())
            .collect()
    };
    for segment in path.split('/') {
        match segment {
            "" | "." => {}
            ".." => {
                if segments.pop().is_none() {
                    return Err(PathError(format!("{path:?} leaves the loader root")));
                }
            }
            other => segments.push(other),
        }
    }
    Ok(segments.join("/"))
}
